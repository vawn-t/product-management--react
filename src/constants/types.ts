export enum Currencies {
  Dollar = '$',
  Euro = '€',
  VND = '₫'
}

export enum ProductTypes {
  Phone = 'Phone',
  Tablet = 'Tablet',
  Laptop = 'Laptop'
}

export const PRODUCT_TYPE_LIST: ProductTypes[] = [
  ProductTypes.Phone,
  ProductTypes.Tablet,
  ProductTypes.Laptop
];

export enum FilterOrderOptions {
  Asc = 'asc',
  Desc = 'desc'
}

export const ORDER_OPTIONS = [FilterOrderOptions.Asc, FilterOrderOptions.Desc];

export type ValidateError = {
  name: string;
  type: string;
  price: string;
  imageUrl: string;
};

export const INIT_ERRORS: ValidateError = {
  name: '',
  price: '',
  type: '',
  imageUrl: ''
};
