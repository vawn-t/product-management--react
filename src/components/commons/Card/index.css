.card {
  display: inline-block;
  background-color: var(--light-gray);
  border: 1px solid var(--white);
  cursor: pointer;
  width: 100%;
  height: 27.75rem;
  max-width: 15.75rem;
  border-radius: 5px;
  text-align: start;
  margin: 1rem;
}

.card:hover {
  border: 1px solid var(--dark);
  box-shadow: inset;
}

.card-body {
  padding: 0.5rem;
  margin: 0 0.75rem;
}

.button-wrapper {
  display: flex;
  justify-content: space-between;
  padding: 0.5rem 0;
}

.card-image {
  border-radius: 5px;
  object-fit: contain;
  overflow: hidden;
  height: 18rem;
}
