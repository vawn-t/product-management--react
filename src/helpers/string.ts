/**
 * Format currency each 3 characters
 *
 * @param price number
 * @returns string
 */
export const currencyFormat = (price: number): string => {
  //convert the string to an array
  const arr = price.toString().split('');

  //back to string with added spaces
  let result = '';
  let count = 0;

  for (let i = arr.length - 1; i >= 0; i--) {
    if (count < 3) {
      result = arr[i] + result;
      count = count + 1;
    } else {
      result = ' ' + result;
      count = 0;
      i = i + 1;
    }
  }

  return result;
};
