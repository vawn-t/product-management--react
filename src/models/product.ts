export interface Product {
  readonly id: string;
  name: string;
  type: string;
  price: number;
  imageUrl: string;
}
