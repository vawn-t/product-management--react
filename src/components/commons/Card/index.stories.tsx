// Libraries
import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

// Components
import Card from './index';

// Styles
import '@assets/styles/App.css';
import '@assets/styles/variables.css';
import '@assets/styles/reset.css';
import './index.css';

// Constants
import { Currencies, ProductTypes } from '@constants/index';

export default {
  title: 'Components/Card',
  component: Card
} as ComponentMeta<typeof Card>;

const Template: ComponentStory<typeof Card> = (args) => <Card {...args} />;

export const Default = Template.bind({});
Default.args = {
  product: {
    id: '1',
    name: 'iPhone 13 Promax',
    imageUrl:
      'https://lh6.googleusercontent.com/mCD1IQ4UfXYJkz6508H5_rspW2kt2HRaTMB8tPQArF8EMSonWBKHCRG0LL47xhI-0V8=w2400',
    price: 14000000,
    type: ProductTypes.Phone
  },
  currency: Currencies.VND,
  onNavigate: () => undefined,
  onDeleteProduct: () => undefined
};
