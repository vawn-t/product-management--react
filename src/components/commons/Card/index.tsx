// Libraries
import React, { memo, useCallback } from 'react';

// Components
import { Image, Title, Button } from '@components/index';

// Constants
import { Currencies } from '@constants/index';

// Helpers
import { currencyFormat } from '@helpers/string';

// Styles
import './index.css';

// Models
import { Product } from '@models/product';

// Types
import { ButtonVariants, FwType, VariantTypes } from '@common-types/index';

export interface CardProps {
  product: Product;
  currency?: Currencies;
  onNavigate: (productId: string) => void;
  onDeleteProduct: (id: string) => void;
}

const Card: React.FC<CardProps> = ({
  product,
  currency = Currencies.VND,
  onNavigate,
  onDeleteProduct
}) => {
  const { name, imageUrl, price, id, type: typeProduct } = product;

  const handleToggleModal = useCallback(() => {
    onNavigate(id);
  }, []);

  const handleDelete = useCallback(() => {
    onDeleteProduct(id);
  }, []);

  return (
    <div className='card'>
      <div className='card-image'>
        <Image imageUrl={imageUrl} alt={name} />
      </div>
      <div className='card-body'>
        <div className='title-wrapper'>
          <Title className='card-title' lineHeight='1.5rem'>
            {name}
          </Title>
        </div>
        <Title
          color='var(--dark)'
          variant={VariantTypes.Subtitle}
          fontStyle='italic'
          lineHeight='2rem'
          fontSize='16px'
        >
          {typeProduct}
        </Title>
        <Title
          variant={VariantTypes.Subtitle}
          fontWeight={FwType.Bold}
          lineHeight='1.5rem'
        >
          {currencyFormat(price)}
          <span> {currency}</span>
        </Title>
        <div className='button-wrapper'>
          <Button onClick={handleToggleModal}>Edit</Button>
          <Button variant={ButtonVariants.Secondary} onClick={handleDelete}>
            Delete
          </Button>
        </div>
      </div>
    </div>
  );
};

export default memo(Card);
