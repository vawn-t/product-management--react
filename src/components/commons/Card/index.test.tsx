// Libraries
import React from 'react';
import { render, fireEvent, cleanup, within } from '@testing-library/react';
import { create } from 'react-test-renderer';
import '@testing-library/jest-dom';

// Components
import Card, { CardProps } from './index';

// Models
import { Product } from '@models/product';

// Images
import { iPhoneImage } from '@assets/index';

// Constants
import { Currencies, ProductTypes } from '@constants/types';

// Helpers
import { currencyFormat } from '@helpers/string';

describe('Card component', () => {
  const mockProduct: Product = {
    id: '999',
    name: 'Product Name',
    price: 1000000,
    imageUrl: iPhoneImage,
    type: ProductTypes.Phone
  };

  const defaultProps: CardProps = {
    product: mockProduct,
    currency: Currencies.VND,
    onNavigate: () => jest.fn(),
    onDeleteProduct: () => jest.fn()
  };

  let container: HTMLElement;
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    container.remove();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly', () => {
    const cardWithVnd = create(<Card {...defaultProps} />).toJSON();
    expect(cardWithVnd).toMatchSnapshot();

    const cardWithEuro = create(
      <Card {...defaultProps} currency={Currencies.Euro} />
    ).toJSON();
    expect(cardWithEuro).toMatchSnapshot();

    const cardWithDollar = create(
      <Card {...defaultProps} currency={Currencies.Dollar} />
    ).toJSON();
    expect(cardWithDollar).toMatchSnapshot();
  });

  test('should render correct amount', () => {
    render(<Card {...defaultProps} />, { container });

    const expectValue = currencyFormat(mockProduct.price);

    const { getByText } = within(container);

    expect(getByText(expectValue)).toBeInTheDocument();
  });

  test('should be open product detail page', () => {
    const myMock = jest.fn();

    const { getByText } = render(
      <Card {...defaultProps} onNavigate={myMock} />,
      { container }
    );

    fireEvent.click(getByText('Edit'));

    expect(myMock.mock.calls.length).toEqual(1);
  });

  test('should be delete product', () => {
    const myMock = jest.fn();

    const { getByText } = render(
      <Card {...defaultProps} onDeleteProduct={myMock} />,
      { container }
    );

    fireEvent.click(getByText('Delete'));

    expect(myMock.mock.calls.length).toEqual(1);
  });
});
