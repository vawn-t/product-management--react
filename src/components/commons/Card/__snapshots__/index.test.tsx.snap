// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should render correctly 1`] = `
<div
  className="card"
>
  <div
    className="card-image"
  >
    <img
      alt="Product Name"
      className="image"
    />
  </div>
  <div
    className="card-body"
  >
    <div
      className="title-wrapper"
    >
      <div
        className="title card-title"
        style={
          Object {
            "color": undefined,
            "fontSize": undefined,
            "fontStyle": "inherit",
            "fontWeight": undefined,
            "lineHeight": "1.5rem",
          }
        }
      >
        Product Name
      </div>
    </div>
    <div
      className="subtitle"
      style={
        Object {
          "color": "var(--dark)",
          "fontSize": "16px",
          "fontStyle": "italic",
          "fontWeight": undefined,
          "lineHeight": "2rem",
        }
      }
    >
      Phone
    </div>
    <div
      className="subtitle"
      style={
        Object {
          "color": undefined,
          "fontSize": undefined,
          "fontStyle": "inherit",
          "fontWeight": 700,
          "lineHeight": "1.5rem",
        }
      }
    >
      1 000 000
      <span>
         
        ₫
      </span>
    </div>
    <div
      className="button-wrapper"
    >
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
      >
        Edit
      </button>
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`Card component should render correctly 2`] = `
<div
  className="card"
>
  <div
    className="card-image"
  >
    <img
      alt="Product Name"
      className="image"
    />
  </div>
  <div
    className="card-body"
  >
    <div
      className="title-wrapper"
    >
      <div
        className="title card-title"
        style={
          Object {
            "color": undefined,
            "fontSize": undefined,
            "fontStyle": "inherit",
            "fontWeight": undefined,
            "lineHeight": "1.5rem",
          }
        }
      >
        Product Name
      </div>
    </div>
    <div
      className="subtitle"
      style={
        Object {
          "color": "var(--dark)",
          "fontSize": "16px",
          "fontStyle": "italic",
          "fontWeight": undefined,
          "lineHeight": "2rem",
        }
      }
    >
      Phone
    </div>
    <div
      className="subtitle"
      style={
        Object {
          "color": undefined,
          "fontSize": undefined,
          "fontStyle": "inherit",
          "fontWeight": 700,
          "lineHeight": "1.5rem",
        }
      }
    >
      1 000 000
      <span>
         
        €
      </span>
    </div>
    <div
      className="button-wrapper"
    >
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
      >
        Edit
      </button>
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`Card component should render correctly 3`] = `
<div
  className="card"
>
  <div
    className="card-image"
  >
    <img
      alt="Product Name"
      className="image"
    />
  </div>
  <div
    className="card-body"
  >
    <div
      className="title-wrapper"
    >
      <div
        className="title card-title"
        style={
          Object {
            "color": undefined,
            "fontSize": undefined,
            "fontStyle": "inherit",
            "fontWeight": undefined,
            "lineHeight": "1.5rem",
          }
        }
      >
        Product Name
      </div>
    </div>
    <div
      className="subtitle"
      style={
        Object {
          "color": "var(--dark)",
          "fontSize": "16px",
          "fontStyle": "italic",
          "fontWeight": undefined,
          "lineHeight": "2rem",
        }
      }
    >
      Phone
    </div>
    <div
      className="subtitle"
      style={
        Object {
          "color": undefined,
          "fontSize": undefined,
          "fontStyle": "inherit",
          "fontWeight": 700,
          "lineHeight": "1.5rem",
        }
      }
    >
      1 000 000
      <span>
         
        $
      </span>
    </div>
    <div
      className="button-wrapper"
    >
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
      >
        Edit
      </button>
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
